core/src/main/java/org/infinispan/interceptors/distribution/BaseDistributionInterceptor.java

```diff
@@ -456,17 +469,15 @@ private void handleMissingResponse(Response response) {
             // Note that keys here are only the subset of keys requested from the node which did not send a valid response
             keys.removeAll(pair.getValue());
             rpcManager.invokeRemotelyAsync(Collections.singleton(pair.getKey()), clusteredGetAllCommand, syncIgnoreLeavers)
-                  .whenComplete(new ClusteredGetAllHandler(pair.getKey(), allFuture, ctx, command, pair.getValue(), contactedNodes));
+                  .whenComplete(new ClusteredGetAllHandler(pair.getKey(), allFuture, ctx, command, pair.getValue(), contactedNodes, remoteGetAllHandler));
          }
          if (!keys.isEmpty()) {
-            // GetAllCommand requires all keys to be wrapped when it comes to execute perform() methods, therefore
-            // we need to remove those for which we have not received any entry
             synchronized (allFuture) {
-               Set<?> strippedKeys = new HashSet<>(allFuture.localCommand.getKeys());
-               strippedKeys.removeAll(keys);
-               // We can't just call command.setKeys() - interceptors might compare keys and actual result set
-               allFuture.localCommand = cf.buildGetAllCommand(strippedKeys, command.getFlagsBitSet(), command.isReturnEntries());
-               allFuture.lostData = true;
+               try {
+                  remoteGetAllHandler.onKeysLost(keys);
+               } catch (Throwable t) {
+                  allFuture.completeExceptionally(t);
```

Comment:
I'm not sure I understand the javadoc of `ClusteredGetAllFuture` either... 

Completing `allFuture` will run some interceptor callbacks synchronously, but not all of them. E.g. the state transfer interceptor callback would suspend the invocation to wait for a new topology, and that would release the `allFuture` monitor, but the new topology could arrive while `remoteGetAllHandler.onKeysLost(keys)` is running, and you'd still have 2 threads accessing the same invocation context.

So I don't think synchronizing `completeExceptionally()` helps, we need to cancel all the `BaseDistributionInterceptor`-related callbacks before completing `allFuture` instead.